File: client/desktop/src/shared/types/machine-state.ts
// Backend MachineState (Core/Models/MachineState.cs) yansıması.
// SignalR /machineHub "StateUpdated" event'i bu şekli gönderir.
// Backend C# pascalCase property'lerini default JSON serializer camelCase'e çevirir.

import type { ErrorCode, MachineMode } from './enums'

export interface MachineState {
  // System status
  mode: MachineMode
  hydraulicMotorState: number // 0=OFF, 1=Starting, 2=Ready
  fanActive: boolean
  alarmActive: boolean
  systemReady: boolean
  safetyBits: number
  errorCode: ErrorCode

  // Pistons
  rightPiston: PistonState
  leftPiston: PistonState
  upperPiston: PistonState
  lowerPiston: PistonState

  // Rotation
  rotation: RotationState

  // Sensors
  sensors: SensorState

  // Pneumatics
  rightPneumatic: PneumaticState
  leftPneumatic: PneumaticState

  // Safety (decoded)
  safety: SafetyState

  // Springback
  springback: SpringbackState

  // SLPIS sensor
  slpisSensorRaw: number
  slpisSensorMm: number
  slpisSensorEffectiveMm: number
  slpisSensorZeroed: boolean

  // Physical state info
  physicalInfo: PhysicalStateInfo

  timestamp: string // ISO-8601 UTC
}

export interface PistonState {
  encoderRaw: number
  positionMm: number
  inPosition: boolean
  moving: boolean
  physicalPositionMm: number
  gonyeOffsetMm: number
  stageOffsetMm: number
  safeBackwardMm: number
  safeForwardMm: number
}

export interface RotationState {
  encoderRaw: number
  activeDirection: number // -1=CCW, 0=Stop, 1=CW
  positionMm: number
  inPosition: boolean
}

export interface SensorState {
  s1PressureBar: number
  s2PressureBar: number
  s1FlowCms: number
  s2FlowCms: number
  oilTempC: number
  oilHumidityPercent: number
  oilLevelPercent: number
}

export interface PneumaticState {
  encoderPosition: number
  activeDirection: number // -1=Back, 0=Stop, 1=Forward
  extended: boolean
  retracted: boolean
}

export interface SafetyState {
  emergencyStopOK: boolean
  motorThermalOK: boolean
  fanThermalOK: boolean
  phaseSequenceOK: boolean
  leftPartSensor: boolean
  rightPartSensor: boolean
  contaminationK1: boolean
  contaminationK2: boolean
  contaminationK3: boolean
}

export interface SpringbackState {
  state: number // 0=Idle .. 99=Error
  detectedPositionMm: number
  pneumaticStartMm: number
  pneumaticEndMm: number
  detectionType: number
  minValueMm: number
  errorCode: number
  isMeasuring: boolean
  isComplete: boolean
  hasError: boolean
}

export interface PhysicalStateInfo {
  isGonyeCompleted: boolean
  currentStage: number
  currentStageName: string | null
}

// Boş initial state — connection kurulmadan önce store'a default olarak konur.
// UI null check yerine bu default'la render eder.
export function emptyMachineState(): MachineState {
  const piston = (): PistonState => ({
    encoderRaw: 0,
    positionMm: 0,
    inPosition: false,
    moving: false,
    physicalPositionMm: 0,
    gonyeOffsetMm: 0,
    stageOffsetMm: 0,
    safeBackwardMm: 0,
    safeForwardMm: 0,
  })
  const pneumatic = (): PneumaticState => ({
    encoderPosition: 0,
    activeDirection: 0,
    extended: false,
    retracted: false,
  })
  return {
    mode: 0,
    hydraulicMotorState: 0,
    fanActive: false,
    alarmActive: false,
    systemReady: false,
    safetyBits: 0,
    errorCode: 0,
    rightPiston: piston(),
    leftPiston: piston(),
    upperPiston: piston(),
    lowerPiston: piston(),
    rotation: {
      encoderRaw: 0,
      activeDirection: 0,
      positionMm: 0,
      inPosition: false,
    },
    sensors: {
      s1PressureBar: 0,
      s2PressureBar: 0,
      s1FlowCms: 0,
      s2FlowCms: 0,
      oilTempC: 0,
      oilHumidityPercent: 0,
      oilLevelPercent: 0,
    },
    rightPneumatic: pneumatic(),
    leftPneumatic: pneumatic(),
    safety: {
      emergencyStopOK: false,
      motorThermalOK: false,
      fanThermalOK: false,
      phaseSequenceOK: false,
      leftPartSensor: false,
      rightPartSensor: false,
      contaminationK1: false,
      contaminationK2: false,
      contaminationK3: false,
    },
    springback: {
      state: 0,
      detectedPositionMm: 0,
      pneumaticStartMm: 0,
      pneumaticEndMm: 0,
      detectionType: 0,
      minValueMm: 0,
      errorCode: 0,
      isMeasuring: false,
      isComplete: false,
      hasError: false,
    },
    slpisSensorRaw: 0,
    slpisSensorMm: 0,
    slpisSensorEffectiveMm: 0,
    slpisSensorZeroed: false,
    physicalInfo: {
      isGonyeCompleted: false,
      currentStage: 0,
      currentStageName: null,
    },
    timestamp: new Date(0).toISOString(),
  }
}

File: engine-dataapi/CncBendingMachine.Core/Models/MachineState.cs
using CncBendingMachine.Core.Enums;

namespace CncBendingMachine.Core.Models;

/// <summary>
/// Complete machine state - pushed to UI via SignalR
/// </summary>
public class MachineState
{
    // System Status
    public MachineMode Mode { get; set; }
    public int HydraulicMotorState { get; set; } // 0=OFF, 1=Starting, 2=Ready
    public bool FanActive { get; set; }
    public bool AlarmActive { get; set; }
    public bool SystemReady { get; set; }
    public int SafetyBits { get; set; }
    public ErrorCode ErrorCode { get; set; }

    // Piston States
    public PistonState RightPiston { get; set; } = new();
    public PistonState LeftPiston { get; set; } = new();
    public PistonState UpperPiston { get; set; } = new();
    public PistonState LowerPiston { get; set; } = new();

    // Rotation State
    public RotationState Rotation { get; set; } = new();

    // Sensor State
    public SensorState Sensors { get; set; } = new();

    // Pneumatic States
    public PneumaticState RightPneumatic { get; set; } = new();
    public PneumaticState LeftPneumatic { get; set; } = new();

    // Safety State (decoded from SafetyBits)
    public SafetyState Safety { get; set; } = new();

    // Springback Measurement State
    public SpringbackState Springback { get; set; } = new();

    // SLPIS Sensor (Geri Esneme Ölçümü)
    public int SlpisSensorRaw { get; set; }
    public double SlpisSensorMm { get; set; }
    public double SlpisSensorEffectiveMm { get; set; }
    public bool SlpisSensorZeroed { get; set; }

    // Physical State Info (Gonye + Stage tracking)
    public PhysicalStateInfo PhysicalInfo { get; set; } = new();

    // Timestamp
    public DateTime Timestamp { get; set; } = DateTime.UtcNow;
}

/// <summary>
/// Physical state information - Gonye and Stage tracking
/// </summary>
public class PhysicalStateInfo
{
    public bool IsGonyeCompleted { get; set; }
    public int CurrentStage { get; set; }
    public string? CurrentStageName { get; set; }
}

public class PistonState
{
    public int EncoderRaw { get; set; }
    public double PositionMm { get; set; }
    public bool InPosition { get; set; }
    public bool Moving { get; set; }

    // Physical position tracking (including Gonye + Stage offsets)
    public double PhysicalPositionMm { get; set; }
    public double GonyeOffsetMm { get; set; }
    public double StageOffsetMm { get; set; }
    public double SafeBackwardMm { get; set; }  // Safe backward limit (encoder value)
    public double SafeForwardMm { get; set; }   // Safe forward limit (encoder value)
}

public class RotationState
{
    public int EncoderRaw { get; set; }
    public int ActiveDirection { get; set; } // -1=CCW, 0=Stop, 1=CW
    public double PositionMm { get; set; }   // Position in mm (PLC sends x100, we divide by 100)
    public bool InPosition { get; set; }
}

public class SensorState
{
    public int S1PressureBar { get; set; }
    public int S2PressureBar { get; set; }
    public int S1FlowCms { get; set; }
    public int S2FlowCms { get; set; }
    public int OilTempC { get; set; }
    public int OilHumidityPercent { get; set; }
    public int OilLevelPercent { get; set; }
}

public class PneumaticState
{
    public int EncoderPosition { get; set; }
    public int ActiveDirection { get; set; } // -1=Back, 0=Stop, 1=Forward
    public bool Extended { get; set; }
    public bool Retracted { get; set; }
}

public class SafetyState
{
    public bool EmergencyStopOK { get; set; }
    public bool MotorThermalOK { get; set; }
    public bool FanThermalOK { get; set; }
    public bool PhaseSequenceOK { get; set; }
    public bool LeftPartSensor { get; set; }
    public bool RightPartSensor { get; set; }
    public bool ContaminationK1 { get; set; }
    public bool ContaminationK2 { get; set; }
    public bool ContaminationK3 { get; set; }
}

/// <summary>
/// Springback (Geri Esneme) measurement state
/// </summary>
public class SpringbackState
{
    /// <summary>
    /// State machine: 0=Idle, 1=PneumaticExtend, 2=Measuring, 3=Detected, 4=PneumaticRetract, 5=Complete, 99=Error
    /// </summary>
    public int State { get; set; }

    /// <summary>
    /// Detected piston position when springback occurred (mm)
    /// </summary>
    public double DetectedPositionMm { get; set; }

    /// <summary>
    /// Pneumatic encoder value at start of measurement (mm)
    /// </summary>
    public double PneumaticStartMm { get; set; }

    /// <summary>
    /// Pneumatic encoder value when springback detected (mm)
    /// </summary>
    public double PneumaticEndMm { get; set; }

    /// <summary>
    /// Detection type: 0=None, 1=Increase, 2=AboveMinimum, 3=Stable
    /// </summary>
    public int DetectionType { get; set; }

    /// <summary>
    /// Minimum pneumatic value seen during measurement (mm)
    /// </summary>
    public double MinValueMm { get; set; }

    /// <summary>
    /// Error code: 0=OK, 1=Timeout, 2=PneumaticFail, 3=Cancelled
    /// </summary>
    public int ErrorCode { get; set; }

    /// <summary>
    /// Helper: Is measurement in progress
    /// </summary>
    public bool IsMeasuring => State >= 1 && State <= 4;

    /// <summary>
    /// Helper: Is measurement complete
    /// </summary>
    public bool IsComplete => State == 5;

    /// <summary>
    /// Helper: Has error
    /// </summary>
    public bool HasError => State == 99;
}
